### PR TITLE
differential rotation now uses isot time instead of custom string.

### PR DESCRIPTION
**Pull request by nabobalis**
See if we compare the iso formatting instead if we avoid the 32bit fail.

Fixes https://github.com/sunpy/sunpy/issues/4559

### CODE REVIEW

**Review by dstansby — CHANGES_REQUESTED** [review]
Although this fixes the tests, I think we should understand what's going on. Could you print `time.jd1` and `time.jd2` for a test that fails so we can see the difference in the underlying time values?

**Comment by nabobalis** [comment]
Had to go around about way to get logging/printing working with xdist.

File: tox.ini [hunk 1]
```diff
@@ -46,6 +46,9 @@ deps =
     setuptools
     # All our tests run in parallel which this plugin allows us to.
     pytest-xdist
+    # Since xdist uses stdout in order for it to work, we have to bypass this and use to
+    # capture any logging we might put in a test for debugging.
```
Comment (nabobalis) [inline tox.ini:50]:
```suggestion
    # xdist uses stdout in order to work, we bypass this to
    # allow logging in tests.
```

**Comment by nabobalis** [comment]
```python


    def test_differential_rotate_observer_full_disk(aia171_test_map):
        # Test a full disk map
        new_observer = get_earth(aia171_test_map.date + 6*u.hr)
        dmap = differential_rotate(aia171_test_map, observer=new_observer)
        assert dmap.data.shape == aia171_test_map.data.shape
        logging.warning(f"{dmap.date.jd1}, {dmap.date.jd2}")
        logging.warning(f"{new_observer.obstime.jd1}, {new_observer.obstime.jd2}")
>       assert dmap.date == new_observer.obstime
E       AssertionError: assert <Time object: scale='utc' format='isot' value=2011-02-15T06:00:00.340> == <Time object: scale='utc' format='isot' value=2011-02-15T06:00:00.340>
E         +<Time object: scale='utc' format='isot' value=2011-02-15T06:00:00.340>
E         -<Time object: scale='utc' format='isot' value=2011-02-15T06:00:00.340>

../../.tox/py38-32bit/lib/python3.8/site-packages/sunpy/physics/tests/test_differential_rotation.py:195: AssertionError
-------------------------------------------------------------------------------- Captured log call ---------------------------------------------------------------------------------
WARNING  sunpy:test_differential_rotation.py:193 2455608.0, -0.24999606481481476
WARNING  sunpy:test_differential_rotation.py:194 2455608.0, -0.24999606481481465
_____________________________________________________________________ test_differential_rotate_time_full_disk ______________________________________________________________________
[gw3] linux -- Python 3.8.6 /project/.tox/py38-32bit/bin/python

aia171_test_map = <sunpy.map.sources.sdo.AIAMap object at 0xe6536190>
SunPy Map
---------
Observatory:		 SDO
Instrument:		 AIA 3
Detecto...0.  ],
       [-0.25,  0.  , -0.5 , ...,  0.75, -0.75,  0.  ],
       [ 0.75,  1.5 , -0.75, ...,  0.  , -0.5 ,  0.5 ]])

    def test_differential_rotate_time_full_disk(aia171_test_map):
        # Test a full disk map
        new_time = aia171_test_map.date + 6*u.hr
        with pytest.warns(UserWarning, match="Using 'time' assumes an Earth-based observer"):
            dmap = differential_rotate(aia171_test_map, time=new_time)
        assert dmap.data.shape == aia171_test_map.data.shape
        # The output map should have the same time as the new time now.
        logging.warning(f"{dmap.date.jd1}, {dmap.date.jd2}")
        logging.warning(f"{new_time.jd1}, {new_time.jd2}")
>       assert dmap.date == new_time
E       AssertionError: assert <Time object: scale='utc' format='isot' value=2011-02-15T06:00:00.340> == <Time object: scale='utc' format='isot' value=2011-02-15T06:00:00.340>
E         +<Time object: scale='utc' format='isot' value=2011-02-15T06:00:00.340>
E         -<Time object: scale='utc' format='isot' value=2011-02-15T06:00:00.340>

../../.tox/py38-32bit/lib/python3.8/site-packages/sunpy/physics/tests/test_differential_rotation.py:244: AssertionError
-------------------------------------------------------------------------------- Captured log call ---------------------------------------------------------------------------------
WARNING  sunpy:test_differential_rotation.py:242 2455608.0, -0.24999606481481476
WARNING  sunpy:test_differential_rotation.py:243 2455608.0, -0.24999606481481465
```

File: tox.ini [hunk 2]
```diff
@@ -46,6 +46,9 @@ deps =
     setuptools
     # All our tests run in parallel which this plugin allows us to.
     pytest-xdist
+    # xdist uses stdout in order to work, we bypass this to
+    # allow logging in tests.
```
Comment (nabobalis) [inline tox.ini:50]:
```suggestion
    # xdist uses stdout in order to work, we bypass this to allow logging in tests.
```

**Comment by dstansby** [comment]
Can you check what the `date-obs` value is in the input and output map metadata? I suspect the error might be something to do with how we save the date in the new map at https://github.com/sunpy/sunpy/blob/56e1ab0c2c992f99e0fe3e6bff468b731a51228c/sunpy/physics/differential_rotation.py#L594

**Comment by nabobalis** [comment]
> Can you check what the `date-obs` value is in the input and output map metadata? I suspect the error might be something to do with how we save the date in the new map at
> 
> https://github.com/sunpy/sunpy/blob/56e1ab0c2c992f99e0fe3e6bff468b731a51228c/sunpy/physics/differential_rotation.py#L594

Locally I get this:
```python
[gw0] PASSED sunpy/physics/tests/test_differential_rotation.py::test_differential_rotate_observer_all_off_disk 
sunpy/physics/tests/test_differential_rotation.py::test_differential_rotate_observer_full_disk WARNING: 2011-02-15T00:00:00.34, 2011-02-15T06:00:00.340 [sunpy.physics.tests.test_differential_rotation]
WARNING: 2455608.0, -0.24999606481481473 [sunpy.physics.tests.test_differential_rotation]
WARNING: 2455608.0, -0.24999606481481473 [sunpy.physics.tests.test_differential_rotation]

[gw0] PASSED sunpy/physics/tests/test_differential_rotation.py::test_differential_rotate_observer_full_disk 
sunpy/physics/tests/test_differential_rotation.py::test_differential_rotate_observer_all_on_disk 
[gw0] PASSED sunpy/physics/tests/test_differential_rotation.py::test_differential_rotate_observer_all_on_disk 
sunpy/physics/tests/test_differential_rotation.py::test_differential_rotate_observer_straddles_limb 
[gw0] PASSED sunpy/physics/tests/test_differential_rotation.py::test_differential_rotate_observer_straddles_limb 
sunpy/physics/tests/test_differential_rotation.py::test_differential_rotate_time_full_disk WARNING: 2011-02-15T00:00:00.34, 2011-02-15T06:00:00.340 [sunpy.physics.tests.test_differential_rotation]
WARNING: 2455608.0, -0.24999606481481473 [sunpy.physics.tests.test_differential_rotation]
```
I will see what the CI gives us.

**Comment by nabobalis** [comment]
```python
===================================================================================== FAILURES =====================================================================================
___________________________________________________________________ test_differential_rotate_observer_full_disk ____________________________________________________________________
[gw3] linux -- Python 3.8.6 /project/.tox/py38-32bit/bin/python

aia171_test_map = <sunpy.map.sources.sdo.AIAMap object at 0xe5f0e3b8>
SunPy Map
---------
Observatory:		 SDO
Instrument:		 AIA 3
Detecto...0.  ],
       [-0.25,  0.  , -0.5 , ...,  0.75, -0.75,  0.  ],
       [ 0.75,  1.5 , -0.75, ...,  0.  , -0.5 ,  0.5 ]])

    def test_differential_rotate_observer_full_disk(aia171_test_map):
        # Test a full disk map
        new_observer = get_earth(aia171_test_map.date + 6*u.hr)
        dmap = differential_rotate(aia171_test_map, observer=new_observer)
        assert dmap.data.shape == aia171_test_map.data.shape
        logging.warning(f"{aia171_test_map.meta['date-obs']}, {dmap.meta['date-obs']}")
        logging.warning(f"{dmap.date.jd1}, {dmap.date.jd2}")
        logging.warning(f"{new_observer.obstime.jd1}, {new_observer.obstime.jd2}")
>       assert dmap.date == new_observer.obstime
E       AssertionError: assert <Time object: scale='utc' format='isot' value=2011-02-15T06:00:00.340> == <Time object: scale='utc' format='isot' value=2011-02-15T06:00:00.340>
E         +<Time object: scale='utc' format='isot' value=2011-02-15T06:00:00.340>
E         -<Time object: scale='utc' format='isot' value=2011-02-15T06:00:00.340>

../../.tox/py38-32bit/lib/python3.8/site-packages/sunpy/physics/tests/test_differential_rotation.py:196: AssertionError
-------------------------------------------------------------------------------- Captured log call ---------------------------------------------------------------------------------
WARNING  sunpy:test_differential_rotation.py:193 2011-02-15T00:00:00.34, 2011-02-15T06:00:00.340
WARNING  sunpy:test_differential_rotation.py:194 2455608.0, -0.24999606481481476
WARNING  sunpy:test_differential_rotation.py:195 2455608.0, -0.24999606481481465
_____________________________________________________________________ test_differential_rotate_time_full_disk ______________________________________________________________________
[gw3] linux -- Python 3.8.6 /project/.tox/py38-32bit/bin/python
```

**Comment by dstansby** [comment]
I had a dive into the FITS standard, and both` 011-02-15T00:00:00.34` and `2011-02-15T06:00:00.340` are valid times. I think the 'correct' fix here would be to change
https://github.com/sunpy/sunpy/blob/56e1ab0c2c992f99e0fe3e6bff468b731a51228c/sunpy/physics/differential_rotation.py#L594
to `.obstime.isot`, and therefore we are relying on astropy to create the correct string (instead of using a custom format string).

**Comment by nabobalis** [comment]
> I had a dive into the FITS standard, and both` 2011-02-15T00:00:00.34` and `2011-02-15T06:00:00.340` are valid times. I think the 'correct' fix here would be to change
> https://github.com/sunpy/sunpy/blob/56e1ab0c2c992f99e0fe3e6bff468b731a51228c/sunpy/physics/differential_rotation.py#L594
> 
> 
> to `.obstime.isot`, and therefore we are relying on astropy to create the correct string (instead of using a custom format string).

I'll do so now.

**Comment by nabobalis** [comment]
Didn't seem to work.

**Comment by dstansby** [comment]
Booooo

**Comment by nabobalis** [comment]
I think the change should be done regardless than a custom string format. But at this point, other than changing the tests I have no ideas.

**Comment by dstansby** [comment]
I continue to be confused, as there's no obvious reason why the dates shouldn't be exactly the same, going to do some more debugging at https://github.com/sunpy/sunpy/pull/4568

**Comment by dstansby** [comment]
I got bored, I think this is a tiny floating point error in round tripping strings and the floating point times, I think comparing the isot strings in the test is fine.

**Comment by nabobalis** [comment]
Will do.

**Comment by Cadair** [comment]
astropy/astropy#10646 would probably be useful here in the future.

**Comment by nabobalis** [comment]
Some of the dev figures have changed. I can change the hash but they look worse now?

**Review by dstansby — APPROVED** [review]

**Comment by dstansby** [comment]
This fixes at least some CI so I'll merge, and take a look at the figure test fails.

**Pull request merged**